util/log: fix the long line splitting behavior with UTF-8 sequences

knz opened the pull request:
Alleviates #61681.
Fixes #64896.

Release note (cli change): Long lines in the `crdb-v2` output
logging format are now split when they exceed 16K, instead
of 1K as previously.

Release note (bug fix): The automatic long line splitting behavior
of the new `crdb-v2` logging format now avoids splitting lines in the
middle of UTF-8 sequences, and avoids unbalanced redaction markers.

knz committed:
util/log: default the line split size to 16K instead of 1K

Release note (cli change): Long lines in the `crdb-v2` output
logging format are now split when they exceed 16K, instead
of 1K as previously.

knz committed:
util/log: fix line splitting behavior

Release note (bug fix): The automatic long line splitting behavior
of the new `crdb-v2` logging format now avoids splitting lines in the
middle of UTF-8 sequences, and avoids unbalanced redaction markers.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64900)
<!-- Reviewable:end -->

bdarnell commented:
I didn't get an email for this PR so I left my comments on the backport #64914.

knz committed:
util/log: better honor the maximum line length

This way if/when we ever make it configurable and document it,
there won't be any surprises.

Release note: None

bdarnell reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rauchenstein)

___
*[pkg/util/log/format_crdb_v2.go, line 424 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64900#-M_LsDCq-XCiAtoSAjDS:-M_LsDCq-XCiAtoSAjDT:b-ndxb95) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b0ff2904d2ba0b2d8f6bd35ea71488d502cdfd/pkg/util/log/format_crdb_v2.go#L424)):*
> ```Go
> 			// If we see an opening redaction marker, remember this fact
> 			// so that we close/open it properly.
> 			if strings.HasPrefix(msg[i:], startRedactionMarker) {
> ```

Yes, the most common case is ascii-only logs. But the worst case (say a giant string of Chinese text) is not implausible, and since we're in logging code that is executed under a global mutex it's good to be paranoid about performance. 

To me, using bytes instead of runes in this loop is what feels like a premature optimization - we're manually duplicating essentially all of the work that the rune loop would do. 

Anyway, since we're under time pressure and don't want to make a more intrusive change this LGTM, just something to consider for the future. 
___
*[pkg/util/log/testdata/crdb_v2_break_lines, line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64900#-M_Lrj8B68EMMPOGcRu6:-M_Lrj8B68EMMPOGcRu7:bxocwbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b0ff2904d2ba0b2d8f6bd35ea71488d502cdfd/pkg/util/log/testdata/crdb_v2_break_lines#L62)):*
> ```
> ☃☃☃☃☃☃☃☃☃☃☃☃
> ----
> I000101 00:00:00.000000 0 :0 ⋮ [-]   ☃☃☃☃☃☃☃☃
> ```

![image.png](https://files.reviewableusercontent.io/images/reviewable/160562/qlj07-XSqPjiSDxihX1hoVoD/image.png)

This is what I see here (for this line and all the snowmen below. The snowmen *above* this point look fine. Anyway, I'm happy to chalk it up to a reviewable/chrome/font oddity. 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @rauchenstein)

___
*[pkg/util/log/format_crdb_v2.go, line 424 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64900#-M_LsDCq-XCiAtoSAjDS:-M_LyJFy-2x2pdz65SGk:bf6azl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b0ff2904d2ba0b2d8f6bd35ea71488d502cdfd/pkg/util/log/format_crdb_v2.go#L424)):*
> since we're in logging code that is executed under a global mutex 

Not any more, no. The critical section starts after the formatting completes.

> we're manually duplicating essentially all of the work that the rune loop would do.

Unfortunately, the logging code does not guarantee that the bytes will be valid UTF-8 (for non-redactable entries); what's the behavior of rune iteration if it finds non-valid UTF-8?
___
*[pkg/util/log/testdata/crdb_v2_break_lines, line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64900#-M_Lrj8B68EMMPOGcRu6:-M_LztgB4EId5m3w9JZ7:bze8ajl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b0ff2904d2ba0b2d8f6bd35ea71488d502cdfd/pkg/util/log/testdata/crdb_v2_break_lines#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

![image.png](https://files.reviewableusercontent.io/images/reviewable/160562/qlj07-XSqPjiSDxihX1hoVoD/image.png)

This is what I see here (for this line and all the snowmen below. The snowmen *above* this point look fine. Anyway, I'm happy to chalk it up to a reviewable/chrome/font oddity. 
</blockquote></details>

yeah that seems to be browser-specific, both in my text editor and in my reviewable I don't see this


<!-- Sent from Reviewable.io -->
